File: golang/215.kth-largest-element-in-an-array.go
/*
 * @lc app=leetcode.cn id=215 lang=golang
 *
 * [215] Kth Largest Element in an Array
 */
package main

import "container/heap"

// @lc code=start
type IntHeap []int

func (h *IntHeap) Less(i, j int) bool {
	return (*h)[i] > (*h)[j]
}

func (h *IntHeap) Swap(i, j int) {
	(*h)[i], (*h)[j] = (*h)[j], (*h)[i]
}

func (h *IntHeap) Len() int {
	return len(*h)
}

func (h *IntHeap) Pop() (v interface{}) {
	*h, v = (*h)[:h.Len()-1], (*h)[h.Len()-1]
	return
}

func (h *IntHeap) Push(v interface{}) {
	*h = append(*h, v.(int))
}
func findKthLargest(nums []int, k int) int {
	h := IntHeap(nums)
	heap.Init(&h)
	ret := 0
	for i := 0; i < k; i++ {
		ret = heap.Pop(&h).(int)
	}
	return ret
}

// @lc code=end

File: cpp/215.kth-largest-element-in-an-array.cpp
/*
 * @lc app=leetcode id=215 lang=cpp
 *
 * [215] Kth Largest Element in an Array
 *
 * https://leetcode.com/problems/kth-largest-element-in-an-array/description/
 *
 * algorithms
 * Medium (68.31%)
 * Likes:    18176
 * Dislikes: 952
 * Total Accepted:    3.2M
 * Total Submissions: 4.7M
 * Testcase Example:  '[3,2,1,5,6,4]\n2'
 *
 * Given an integer array nums and an integer k, return the k^th largest
 * element in the array.
 *
 * Note that it is the k^th largest element in the sorted order, not the k^th
 * distinct element.
 *
 * Can you solve it without sorting?
 *
 *
 * Example 1:
 * Input: nums = [3,2,1,5,6,4], k = 2
 * Output: 5
 * Example 2:
 * Input: nums = [3,2,3,1,2,4,5,5,6], k = 4
 * Output: 4
 *
 *
 * Constraints:
 *
 *
 * 1 <= k <= nums.length <= 10^5
 * -10^4 <= nums[i] <= 10^4
 *
 *
 */

#include "leetcode.h"
// @lc code=start
class Solution {
public:
  int findKthLargest(vector<int> &nums, int k) {
    return findKthLargestHeap(nums, k);
  }
  int findKthLargesSort(vector<int> &nums, int k) {
    sort(nums.begin(), nums.end());
    return nums[nums.size() - k];
  }

  int findKthLargestHeap(vector<int> &nums, int k) {
    make_heap(std::begin(nums),std::end(nums),std::less());

    for (int i = 0; i < k-1; i++) {
        pop_heap( std::begin(nums),std::end(nums),std::less());
        nums.pop_back();
    }
    pop_heap(std::begin(nums), std::end(nums), std::less());
    return nums.back();
  }
  int findKthLargestQuickSelect(vector<int> &nums, int k) {
    
  }
  };
  // @lc code=end
